### PR TITLE
Auto update ticket quantity when the datetime capacity is changed

### LINKED ISSUE #1747: Automatically decrement ticket quantity fields when related datetimes reg limit gets set
This was a feature in the original ticket editor where the value for the registration limit for a datetime would automatically "sync" the related ticket's quantity field. Animated gif to show the ticket fields updating after saving the event's changes:

ticket-datetime-limit

In the new editor, if you change the datetime reg limit to a smaller number, the related ticket quantity field does not automatically update.
Screen Shot 2019-10-28 at 5 31 21 PM

### PR DESCRIPTION
**Pull request by manzoorwanijk**
Fixes #1747 

## How has this been tested
monkey testing

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js [hunk 1]
```diff
@@ -131,11 +133,13 @@ const useDateEntityInputConfig = ( {
 			label: __( 'Capacity', 'event_espresso' ),
 			default: -1,
 			changeListener: ( value ) => {
-				dateEntity.regLimit = parseInfinity(
+				const regLimit = parseInfinity(
 					value || -1,
 					true,
 					true
 				);
+				dateEntity.regLimit = regLimit;
```
Comment (nicolad) [inline assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js]: Although this is not directly related to this PR.
We need to reconsider how state management is done.
One way of improvement is by using immutability.

Comment (manzoorwanijk) [reply assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js]: It's already immutable :)

**Comment by manzoorwanijk** [comment]
There is a memoization issue though when you update the datetime capacity inline, the change doesn't reflect unless there is a change in the UI filters. I think that can be fixed in another PR.

File: assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js (hunk 1 above)
Comment (tn3rb) [reply assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js]: > It's already immutable

???

The model entities are not immutable, some fields that use value objects like date or money fields are immutable, but the objects in whole are not.

or are you referring to the state object itself? In which case, yes the state is immutable.

> We need to reconsider how state management is done.
> One way of improvement might be: immutability.

How would that work using the code above as an example? 
Are you suggesting constructing a whole new event date object when the capacity (regLimit) property changes value?

File: assets/src/hooks/use-event-date-update-related-tickets.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+const useEventDateUpdateRelatedTickets = ( eventDate ) => {
+	const { tickets: relatedTickets, ticketsLoaded } = useEventDateTickets( eventDate );
+	return useCallback( ( { capacity = null } ) => {
+		if ( ticketsLoaded ) {
+			// Make sure that the ticket qty value is compared with
+			// a non negative datetime capacity value in Math.min()
+			const nonNegativeDTTCapacity = parseInfinity( capacity, false, false );
+			relatedTickets.forEach( ( ticket ) => {
+				if ( capacity !== null ) {
```
Comment (tn3rb) [inline assets/src/hooks/use-event-date-update-related-tickets.js]: since we don't need to do anything when capacity is null, then why not merge this with the conditional above and then bypass everything?

ie:

```js
if ( ticketsLoaded && capacity !== null ) {
    // loop through tickets and possibly adjust quantity
}
```

Comment (manzoorwanijk) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: `parseInfinity` already converts `null` to `Infinity` or `-1` by making use of `isNil( value );`

Comment (tn3rb) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: My comment has nothing to do with `parseInfinity` or how it converts values.

Look at the entire function again:


```js
	( { capacity = null } ) => {
		if ( ticketsLoaded ) {
			// Make sure that the ticket qty value is compared with
			// a non negative datetime capacity value in Math.min()
			const nonNegativeDTTCapacity = parseInfinity( capacity, false, false );
			relatedTickets.forEach( ( ticket ) => {
				if ( capacity !== null ) { <<<--- WE ARE CHECKING THIS REPEATEDLY
					const nonNegativeTKTQty = parseInfinity( ticket.qty, false, false );
					const qty = Math.min( nonNegativeDTTCapacity, nonNegativeTKTQty );
					ticket.qty = parseInfinity( qty, true, true );
				}
			} );
		}
	}
```

`capacity` is passed into the callback but is not modified in any way before being compared against null while looping through tickets to decide whether to adjust the ticket qty. But since the value of `capacity` is never changed anywhere in the function, and we don't end up doing anything when it's incoming value is `null`, we are running that same comparison over and over. We could move that comparison into the very first conditional and then bypass **everything** when `capacity` is `null` since we are not going to end up doing anything anyways.

```js
if ( ticketsLoaded && capacity !== null ) { <<<--- ONLY NEED TO CHECK CAP ONCE
	// Make sure that the ticket qty value is compared with
	// a non negative datetime capacity value in Math.min()
	const nonNegativeDTTCapacity = parseInfinity( capacity, false, false );
	relatedTickets.forEach( ( ticket ) => {
		const nonNegativeTKTQty = parseInfinity( ticket.qty, false, false );
		const qty = Math.min( nonNegativeDTTCapacity, nonNegativeTKTQty );
		ticket.qty = parseInfinity( qty, true, true );
	} );
}
```

File: assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js (hunk 1 above)
Comment (nicolad) [reply assets/src/editor/events/dates-and-times/editor-date/edit-form/use-date-entity-input-config.js]: `Are you suggesting constructing a whole new event date object when the capacity (regLimit) property changes value?`
Yes, that would be faster since old object will be garbage collected and memory got freed.
From what I saw we already use that pattern since we are using `@wordpress/data`

File: assets/src/hooks/use-event-date-update-related-tickets.js (hunk 2 above)
Comment (manzoorwanijk) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: I was pointing towards how the value actually reaches there. If you check the `onChange` handler for `capacity`,
```JS
onChange: ( cap ) => {
	cap = parseInfinity( cap, true, true );
	eventDate.regLimit = cap;
	updateRelatedTickets( { capacity: cap } );
	return <InfinitySymbol value={ cap } asInt />;
}
```
You can see that the value is not null.
May be we can pass the actual value of `cap` from the `onChange` handler to the hook for updating the ticket quantity.

Comment (manzoorwanijk) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: Why I have used `if ( capacity !== null )`?
As I mentioned in the commit message, it is a generic hook which can/may be used to update other fields in the related tickets. This is also the reason that the arguments are passed as object `( { capacity = null } )` instead of `( capacity = null )`.

File: assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js [hunk 3]
```diff
@@ -7,61 +7,66 @@ import { parseInfinity } from '@eventespresso/utils';
...
-						return <InfinitySymbol value={ cap } asInt />;
+const EventDateDetailsPanel = ( { eventDate } ) => {
+	const updateRelatedTickets = useEventDateUpdateRelatedTickets( eventDate );
+	return useMemo(
+		() => {
+			const details = [
+				{
+					id: 'ee-event-date-sold',
+					label: __( 'sold', 'event_espresso' ),
+					value: eventDate.sold || 0,
```
Comment (nicolad) [inline assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: might `eventDate` object be null ?
if so, then maybe smth like: `value: (eventDate && eventDate.sold) || 0`

File: assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js [hunk 4]
```diff
@@ -7,61 +7,66 @@ import { parseInfinity } from '@eventespresso/utils';
...
+					label: __( 'capacity', 'event_espresso' ),
+					value: <InfinitySymbol value={ eventDate.regLimit } asInt />,
+					editable: {
+						type: 'text',
+						valueType: 'infinite',
+						onChange: ( cap ) => {
+							cap = parseInfinity( cap, true, true );
+							eventDate.regLimit = cap;
+							updateRelatedTickets( { capacity: cap } );
+							return <InfinitySymbol value={ cap } asInt />;
```
Comment (nicolad) [inline assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: What about:
```
const capacity = parseInfinity( cap, true, true );
eventDate.regLimit = cap;
updateRelatedTickets( { capacity } );
return <InfinitySymbol value={ cap } asInt />;
```
this way we don't mutate cap

File: assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js (hunk 3 above)
Comment (manzoorwanijk) [reply assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: `evenDate` is expected to be non null.

File: assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js (hunk 4 above)
Comment (manzoorwanijk) [reply assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: Yes it makes the code better. Thanks

Comment (manzoorwanijk) [reply assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: Done :)

Comment (nicolad) [reply assets/src/editor/events/dates-and-times/editor-date/grid-view/event-date-details-panel.js]: Thanks :bowing_man:

File: assets/src/hooks/use-event-date-update-related-tickets.js (hunk 2 above)
Comment (manzoorwanijk) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: > 
> 
> Why I have used `if ( capacity !== null )`?
> As I mentioned in the commit message, it is a generic hook which can/may be used to update other fields in the related tickets. This is also the reason that the arguments are passed as object `( { capacity = null } )` instead of `( capacity = null )`.

@tn3rb I understand your point, please read this quoted comment again. What if in future we want to update something other than quantity for the related tickets? We could use the same hook and that time capacity may not need to be changed. I hope I am not trying create an ultra hypothetical use case. 😄

Comment (tn3rb) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: **GO FOR IT!!!**  😆

Comment (manzoorwanijk) [reply assets/src/hooks/use-event-date-update-related-tickets.js]: Done! Please have a look 😊
